### ISSUE TITLE
CancellationToken is not honored for sync methods

### ISSUE THREAD

**Issue by AArnott** [Microsoft/vs-streamjsonrpc#86]
Because the reader loop executes methods directly within the loop, any sync methods (or async methods before their first yielding _await_) will never recognize a cancellation request from the client if they were to take a `CancellationToken`.

We could fix this by invoking methods using `Task.Run` so that they can't block our reader loop, thus freeing it up to receive the cancellation request. This _might_ break any servers that depend on their sync methods being executed sequentially or in-order since the change would allow multiple sync server methods to execute concurrently.

**Design proposal**

Add a `SynchronizationContext` field to the `JsonRpc` class and all local method invocations will be done by calling `SynchronizationContext.Post`. This `SynchronizationContext` will be either the default one (which schedules work concurrently to the threadpool) or one that is "captured" or otherwise provided in the constructor. That way, folks who want to preserve order of incoming messages can do so by supplying a single-threaded `SynchronizationContext`, but otherwise execution becomes concurrent.
This design will allow even the single threaded servers to observe cancellation requests while sync methods are executing.

CC: @BertanAygun @heejaechang @tinaschrepfer @AlexEyler

**Comment by heejaechang** [comment]
does this imply JsonRpc was not thread-safe before? (2 threads calling 2 methods) or it was always serialized for sync method (method 1 must exist before method 2 run)

I think I had wrong impression before, then. I thought if 2 threads calls 2 different sync methods, it would run in parallel in server side. it is just message that got serialized not actual method run.

for me, new behavior seems more inline with what I have expected.

**Comment by AArnott** [comment]
> does this imply JsonRpc was not thread-safe before?

No, StreamJsonRpc is perfectly thread safe as far as we know. But it did not execute server methods concurrently (except that async methods can resume execution after yielding whenever they want).

> I think I had wrong impression before, then. I thought if 2 threads calls 2 different sync methods, it would run in parallel in server side. it is just message that got serialized not actual method run.

So did I. I only discovered this yesterday when verifying the behavior for @BertanAygun who actually wants the opposite.

**Comment by heejaechang** [comment]
@AArnott @BertanAygun just to clarify.

if caller does this

await rpc.InvokeAsync("SyncMethod1", ...);
await rpc.InvokeAsync("SyncMethod2", ...);

then SyncMethod1 and 2 runs sequentially.

but question is, if user does

/* no await here */
rpc.InvokeAsync("SyncMethod1", ...);
rpc.InvokeAsync("SyncMethod2", ...);

should SyncMethod2 only run after SyncMethod1 or not? is that question?

then my question is that, since rpc.InvokeAsync is not awaited, does JsonRpc guarantee that call message to SyncMethod1 is delivered before SyntaxMethod2? assuming InvokeAsync internally use task/await I don't believe there is guarantee which task will run first in VS side even if it is blocked on server side?

...

anyway, regardless of how it works. I would prefer its behavior to be same as any normal C# async method.

if I do this
var task1 = processor.ProcessAsync(...);
var task2 = processor.ProcessAsync(...);

I would expect task1 and task2 to run concurrently. so I would expect same for 
rpc.InvokeAsync("SyncMethod1", ...);
rpc.InvokeAsync("SyncMethod2", ...);

it invoking sync method in server won't matter since I am calling through InvokeAsync in call site.

if one wants it to behavior like SyncMethod, I think caller should either do

await rpc.InvokeAsync(...) or use Wait or GetAwaiter().GetResult()

**Comment by AArnott** [comment]
> then my question is that, since rpc.InvokeAsync is not awaited, does JsonRpc guarantee that call message to SyncMethod1 is delivered before SyntaxMethod2?

Yes: JsonRpc will always transmit messages in the order the calls are made by the client. And the remote side always invokes methods on the server in the order they are read from the stream. 

Sync behavior at the client is out of scope of this issue. The behavior in question is how the _server side_ works (whether it continues to listen for and dispatch messages even while a prior message dispatch is still in progress). And yes, I agree with you on all your expectations. I was surprised too when I realized the current behavior. This issue tracks making it match what you and I expected.

I edited the original issue description to include a design proposal for the fix.

**Comment by AArnott** [comment]
I misread the code. This is Not Repro. There actually *is* code to ensure we can keep executing new server methods even before prior ones have returned. I'm adding a test to prove it.

So @BertanAygun this means order of execution of server methods is *not* guaranteed, since they can actually run in parallel.
